corehq/messaging/scheduling/models/content.py

```diff
@@ -478,3 +482,111 @@ def send(self, recipient, logged_event, phone_entry=None):
+        renderer = self.get_template_renderer(recipient)
+        return renderer.render(subject), renderer.render(message)
+
+    def build_fcm_data_field(self, recipient):
+        data = {}
+        if self.action:
+            data = {
+                'action': self.action,
+                'username': recipient.raw_username,
+                'domain': recipient.domain,
+                'created_at': datetime.now(timezone.utc).isoformat()
+            }
+        return data
+
+    def send(self, recipient, logged_event, phone_entry=None):
+        domain_obj = Domain.get_by_name(logged_event.domain)
+
+        logged_subevent = logged_event.create_subevent_from_contact_and_content(
+            recipient,
+            self,
+            case_id=self.case.case_id if self.case else None,
+        )
+        subject = message = data = None
+
+        if not settings.FCM_CREDS:
+            logged_subevent.error(MessagingEvent.ERROR_FCM_NOT_AVAILABLE)
+            return
+
+        if not isinstance(recipient, CommCareUser):
+            logged_subevent.error(MessagingEvent.ERROR_FCM_UNSUPPORTED_RECIPIENT)
+            return
+
+        if self.message_type == self.MESSAGE_TYPE_NOTIFICATION:
+            if not (self.subject or self.message):
+                logged_subevent.error(MessagingEvent.ERROR_NO_MESSAGE)
+                return
+
+            recipient_language_code = recipient.get_language_code()
+            subject = self.get_translation_from_message_dict(
+                domain_obj,
+                self.subject,
+                recipient_language_code
+            )
+
+            message = self.get_translation_from_message_dict(
+                domain_obj,
+                self.message,
+                recipient_language_code
+            )
+
+            try:
+                subject, message = self.render_subject_and_message(subject, message, recipient)
+            except Exception:
+                logged_subevent.error(MessagingEvent.ERROR_CANNOT_RENDER_MESSAGE)
+                return
+        else:
+            if not self.action:
+                logged_subevent.error(MessagingEvent.ERROR_NO_MESSAGE)
+                return
+            data = self.build_fcm_data_field(recipient)
+
+        try:
+            devices_fcm_tokens = self.get_recipient_devices_fcm_tokens(recipient)
+        except FCMTokenValidationException as e:
+            logged_subevent.error(e.error_type, additional_error_text=e.additional_text)
+            return
+
+        FCMUtil().send_to_multiple_devices(registration_tokens=devices_fcm_tokens, title=subject, body=message,
+                                           data=data)
+        logged_subevent.completed()
+
+    def get_recipient_devices_fcm_tokens(self, recipient):
+        devices_fcm_tokens = recipient.get_devices_fcm_tokens()
+        if not devices_fcm_tokens:
+            raise FCMTokenValidationException(MessagingEvent.ERROR_NO_FCM_TOKENS)
```

Comment:
I am worried these could be a huge number of unnecessary logs when someone would initiate a message to users and a lot of them might not have fcm token sent to HQ. Can we move this check up when a message is even initiated to not initiate it at all and avoid all this logging?